terraform: add the version component to terraform provider paths

siers opened the pull request:
Terraform checks the provider versions, but this breaks if the versions
are not provided, as they can be, if the plugins are provided by nix.

###### Motivation for this change

Fixes #34423.

I'm not sure whether a [quick-fix](https://github.com/siers/nixpkgs/commits/terraform-provider-symlink-versioning-quickfix)(see first commit) would be more appropriate here or not.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"` **— no, just a couple**
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

copumpkin commented:
I think this looks good, thanks! @zimbatm any thoughts?

zimbatm commented:
I don't get it. Couldn't the plugins be built with the right output format directly?

copumpkin commented:
Oh yeah, that's probably better

siers commented:
As mentioned in the initial description of the PR, this might work then: https://github.com/siers/nixpkgs/commit/4ce469e74c0d13954ffd10d9fe282b39518e3d93.

1. I don't know any other way to change the output binary name, since `buildFlags` also get passed to non-`go build` commands, so I just `mv`ed it. 2. I also was a little prematurely worried that something else might try to look for the binary of the expected name and not find it, but one should probably wait for that to happen before getting worried.

zimbatm commented:
yeah let's do https://github.com/siers/nixpkgs/commit/4ce469e74c0d13954ffd10d9fe282b39518e3d93 with a comment explaining the reason for the rename

siers committed:
terraform: add the version component to terraform provider paths

Terraform checks the provider versions, but this breaks if the versions
are not provided, as they can be, if the plugins are provided by nix.

siers commented:
I've added a comment in both source and commit message, here's the commit https://github.com/siers/nixpkgs/commit/53cb2b1cd8d349d279c1856351dacd75f1dfc5cc. Should I create a new PR or…?

zimbatm commented:
it's fine to force-push and replace the content of this branch if you want. or create a new PR and @mention me

siers commented:
@zimbatm Hey, I've forced the branch.

zimbatm commented:
@GrahamcOfBorg build terraform-full

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34497&attempt_id=bec73ade-4ec7-486f-b3f1-eb7a1f9c6d39)

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0/share/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/golang/protobuf/ptypes/regen.sh: interpreter directive changed from "/bin/bash -e" to "/nix/store/7fxbh1yhagvwbdrmdyyy5ghcjhwjndhs-bash-4.4-p12/bin/bash -e"
/nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0/share/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/posener/complete/test.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/7fxbh1yhagvwbdrmdyyy5ghcjhwjndhs-bash-4.4-p12/bin/bash"
/nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0/share/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/golang.org/x/sys/unix/mkall.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/7fxbh1yhagvwbdrmdyyy5ghcjhwjndhs-bash-4.4-p12/bin/bash"
/nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0/share/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/golang.org/x/sys/unix/mkerrors.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/7fxbh1yhagvwbdrmdyyy5ghcjhwjndhs-bash-4.4-p12/bin/bash"
/nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0/share/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/codegen.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/7fxbh1yhagvwbdrmdyyy5ghcjhwjndhs-bash-4.4-p12/bin/bash"
/nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0/share/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/coverage.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/7fxbh1yhagvwbdrmdyyy5ghcjhwjndhs-bash-4.4-p12/bin/bash"
checking for references to /build in /nix/store/xf987pqr85b49mgwdz3bd7z49szzhk96-terraform-provider-azurerm-1.0.0...
cannot build derivation '/nix/store/hd1yj1l7pa82v53ryf2ahp2km5iqpps0-tf-plugin-env.drv': 30 dependencies couldn't be built
cannot build derivation '/nix/store/w1xqbckws4bwn2cilwyp0b71445idcl1-terraform-0.11.3-with-plugins.drv': 1 dependencies couldn't be built
error: build of '/nix/store/w1xqbckws4bwn2cilwyp0b71445idcl1-terraform-0.11.3-with-plugins.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34497&attempt_id=none)

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/6h6va7h33q9hvcg9b4w6z75b4dcrv3hj-terraform-provider-vsphere-1.1.1/share/go/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/google.golang.org/grpc/codegen.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/yq03c2ny43mc24j7dq5riznzb09ddhpq-bash-4.4-p12/bin/bash"
/nix/store/6h6va7h33q9hvcg9b4w6z75b4dcrv3hj-terraform-provider-vsphere-1.1.1/share/go/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/google.golang.org/grpc/install-protobuf.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/yq03c2ny43mc24j7dq5riznzb09ddhpq-bash-4.4-p12/bin/bash"
/nix/store/6h6va7h33q9hvcg9b4w6z75b4dcrv3hj-terraform-provider-vsphere-1.1.1/share/go/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/golang.org/x/sys/unix/mkall.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/yq03c2ny43mc24j7dq5riznzb09ddhpq-bash-4.4-p12/bin/bash"
/nix/store/6h6va7h33q9hvcg9b4w6z75b4dcrv3hj-terraform-provider-vsphere-1.1.1/share/go/src/github.com/terraform-providers/terraform-provider-vsphere/vendor/golang.org/x/sys/unix/mkerrors.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/yq03c2ny43mc24j7dq5riznzb09ddhpq-bash-4.4-p12/bin/bash"
checking for references to /tmp/nix-build-terraform-provider-vsphere-1.1.1.drv-0 in /nix/store/6h6va7h33q9hvcg9b4w6z75b4dcrv3hj-terraform-provider-vsphere-1.1.1...
building path(s) ‘/nix/store/cwdmzz96xgdjrgc5j95pxv1zq1a84hlr-tf-plugin-env’
Use of uninitialized value $path in -e at /nix/store/9g4wc31j7a2xp22xpgwr0qssfxahxdzl-builder.pl line 172.
created 73 symlinks in user environment
building path(s) ‘/nix/store/ixmz361hklgzsnb36dksn15v8rwaijhz-terraform-0.11.3-with-plugins’
/nix/store/ixmz361hklgzsnb36dksn15v8rwaijhz-terraform-0.11.3-with-plugins
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34497&attempt_id=40f6fe04-4691-40d4-9747-3f5356fd339b)

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/353ml9fm6nlxbi83vpdwmwv9h4rzf2xx-terraform-provider-triton-0.4.1/share/go/src/github.com/terraform-providers/terraform-provider-triton/vendor/golang.org/x/sys/unix/mkall.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/s8mff1kmnc63b21ybdid2ni0fw7mzy7r-bash-4.4-p12/bin/bash"
/nix/store/353ml9fm6nlxbi83vpdwmwv9h4rzf2xx-terraform-provider-triton-0.4.1/share/go/src/github.com/terraform-providers/terraform-provider-triton/vendor/golang.org/x/sys/unix/mkerrors.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/s8mff1kmnc63b21ybdid2ni0fw7mzy7r-bash-4.4-p12/bin/bash"
/nix/store/353ml9fm6nlxbi83vpdwmwv9h4rzf2xx-terraform-provider-triton-0.4.1/share/go/src/github.com/terraform-providers/terraform-provider-triton/vendor/google.golang.org/genproto/regen.sh: interpreter directive changed from "/bin/bash" to "/nix/store/s8mff1kmnc63b21ybdid2ni0fw7mzy7r-bash-4.4-p12/bin/bash"
/nix/store/353ml9fm6nlxbi83vpdwmwv9h4rzf2xx-terraform-provider-triton-0.4.1/share/go/src/github.com/terraform-providers/terraform-provider-triton/vendor/google.golang.org/grpc/codegen.sh: interpreter directive changed from "/usr/bin/env bash" to "/nix/store/s8mff1kmnc63b21ybdid2ni0fw7mzy7r-bash-4.4-p12/bin/bash"
/nix/store/353ml9fm6nlxbi83vpdwmwv9h4rzf2xx-terraform-provider-triton-0.4.1/share/go/src/github.com/terraform-providers/terraform-provider-triton/vendor/google.golang.org/grpc/vet.sh: interpreter directive changed from "/bin/bash" to "/nix/store/s8mff1kmnc63b21ybdid2ni0fw7mzy7r-bash-4.4-p12/bin/bash"
building '/nix/store/876n5qn03xq9qrj1bcnhjmj9pbqdjyly-tf-plugin-env.drv'...
Use of uninitialized value $path in -e at /nix/store/9g4wc31j7a2xp22xpgwr0qssfxahxdzl-builder.pl line 172.
created 73 symlinks in user environment
building '/nix/store/m3y4vqiw3c2fzhb0xm27v1ylqdklal8z-terraform-0.11.3-with-plugins.drv'...
/nix/store/1cjwk79x5b0q8n9sa6bhgkji66ah6xl2-terraform-0.11.3-with-plugins
```
</p></details>

siers commented:
Hi @zimbatm, maybe you want to merge this onto the 17.09? It also has the `-full` packages, so it would surely benefit from this.

zimbatm commented:
pushed as b2e7bb5ee80303ec6d0671085f5da424014a601f
